src/Nonlinear/ReverseAD/types.jl

```diff
@@ -0,0 +1,185 @@
+    seed_matrix::Matrix{Float64}
+    linearity::Linearity
+    # subexpressions which this function depends on, ordered for forward pass.
+    dependent_subexpressions::Vector{Int}
+
+    function _FunctionStorage(
+        nodes::Vector{Nonlinear.Node},
+        const_values,
+        num_variables,
+        coloring_storage::Coloring.IndexedSet,
+        want_hess::Bool,
+        subexpressions::Vector{_SubexpressionStorage},
+        dependent_subexpressions,
+        subexpression_linearity,
+        subexpression_edgelist,
+        subexpression_variables,
+        moi_index_to_consecutive_index,
+    )
+        nodes = _replace_moi_variables(nodes, moi_index_to_consecutive_index)
+        adj = Nonlinear.adjacency_matrix(nodes)
+        N = length(nodes)
+        empty!(coloring_storage)
+        _compute_gradient_sparsity!(coloring_storage, nodes)
+        for k in dependent_subexpressions
+            _compute_gradient_sparsity!(
+                coloring_storage,
+                subexpressions[k].nodes,
+            )
+        end
+        grad_sparsity = sort!(collect(coloring_storage))
+        empty!(coloring_storage)
+        if want_hess
+            linearity = _classify_linearity(nodes, adj, subexpression_linearity)
+            edgelist = _compute_hessian_sparsity(
+                nodes,
+                adj,
+                linearity,
+                coloring_storage,
+                subexpression_edgelist,
+                subexpression_variables,
+            )
+            hess_I, hess_J, rinfo = Coloring.hessian_color_preprocess(
+                edgelist,
+                num_variables,
+                coloring_storage,
+            )
+            seed_matrix = Coloring.seed_matrix(rinfo)
+            return new(
+                nodes,
+                adj,
+                const_values,
+                zeros(N),  # forward_storage,
+                zeros(N),  # partials_storage,
+                zeros(N),  # reverse_storage,
+                grad_sparsity,
+                hess_I,
+                hess_J,
+                rinfo,
+                seed_matrix,
+                linearity[1],
+                dependent_subexpressions,
+            )
+        else
+            return new(
+                nodes,
+                adj,
+                const_values,
+                zeros(N),  # forward_storage,
+                zeros(N),  # partials_storage,
+                zeros(N),  # reverse_storage,
+                grad_sparsity,
+                Int[],
+                Int[],
+                Coloring.RecoveryInfo(),
+                Array{Float64}(undef, 0, 0),
+                NONLINEAR,
+                dependent_subexpressions,
+            )
+        end
+    end
+end
+
+"""
+    NLPEvaluator(
+        data::Nonlinear.NonlinearData,
+        ordered_variables::Vector{MOI.VariableIndex},
+    )
+
+Return an `MOI.AbstractNLPEvaluator`. Before using, you must initialize the
```

Comment:
How does the constructor for a concrete type return `MOI.AbstractNLPEvaluator`? We can directly specify that it returns  `NLPEvaluator`.

To document the type, something more like:
`Implements the MOI.AbstractNLPEvaluator interface`

Reply:
Done.